src/mds/nameserver2/writerlock.cpp

```diff
@@ -0,0 +1,170 @@
+/*
+ *  Copyright (c) 2022 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: Curve
+ * Created Date: 2022-07-25
+ * Author: YangFan (fansehep)
+ */
+
+#include "src/mds/nameserver2/writerlock.h"
+#include <bthread/bthread.h>
+#include <ostream>
+#include <functional>
+#include <chrono>
+#include "proto/nameserver2.pb.h"
+#include "src/common/encode.h"
+
+
+namespace curve {
+namespace mds {
+
+
+using ::curve::common::TimeUtility;
+using ::curve::common::NameLockGuard;
+using ::curve::common::EncodeBigEndian;
+using ::curve::common::TimeUtility;
+
+const char WRITER_LOCK_PREFIX[] = "WriterLock_";
+const uint64_t WRITER_LOCK_PREFIX_SIZE = 12;
+
+inline size_t WriterLock::Hash(const std::string& key) {
+    return std::hash<std::string>{}(key);
+}
+
+std::string WriterLock::EncodeKey(const std::string& writerinfo) {
+    std::string key = WRITER_LOCK_PREFIX;
+    auto prefix_len = WRITER_LOCK_PREFIX_SIZE;
+    uint64_t num = Hash(writerinfo);
+    key.resize(prefix_len + sizeof(uint64_t));
+    EncodeBigEndian(&(key[prefix_len]), num);
+    return key;
```

Comment:
I think the key should be encoded by a unique prefix and file info like file id.